overlord/devicestate/devicestate_serial_test.go

```diff
@@ -2408,3 +2409,193 @@ func (s *deviceMgrSerialSuite) TestDeviceSerialRestoreHappy(c *C) {
+			gadgetName:          "gadget",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "",
+			expected:            true,
+		},
+		{
+			storeAccess:         "offline",
+			gadgetName:          "gadget",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "https://example.com",
+			expected:            true,
+		},
+		{
+			storeAccess:         "offline",
+			gadgetName:          "gadget",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "",
+			expected:            false,
+		},
+		{
+			storeAccess:         "offline",
+			gadgetName:          "gadget",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "https://example.com",
+			expected:            true,
+		},
+	}
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	for _, t := range testCases {
+		tr := config.NewTransaction(s.state)
+		err := tr.Set("core", "store.access", t.storeAccess)
+		c.Assert(err, IsNil)
+
+		if t.gadgetName != "" {
+			err = tr.Set(t.gadgetName, "device-service.access", t.deviceServiceAccess)
+			c.Assert(err, IsNil)
+			err = tr.Set(t.gadgetName, "device-service.url", t.deviceServiceURL)
+			c.Assert(err, IsNil)
+		}
+		tr.Commit()
+
+		shouldReqest, err := devicestate.ShouldRequestSerial(s.state, t.gadgetName)
+		c.Check(err, IsNil)
+		c.Check(shouldReqest, Equals, t.expected, Commentf("%+v", t))
+	}
+}
+
+func (s *deviceMgrSerialSuite) TestDeviceManagerFullAccess(c *C) {
+	s.state.Lock()
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Brand: "canonical",
+		Model: "pc",
+	})
+
+	s.makeModelAssertionInState(c, "canonical", "pc", map[string]interface{}{
+		"architecture": "amd64",
+		"kernel":       "pc-kernel",
+		"gadget":       "pc",
+	})
+
+	devicestatetest.MockGadget(c, s.state, "pc", snap.R(2), nil)
+	s.state.Set("seeded", true)
+	s.state.Unlock()
+
+	err := s.mgr.Ensure()
+	c.Assert(err, IsNil)
+
+	s.state.Lock()
+	change := s.findBecomeOperationalChange()
+	tasks := change.Tasks()
+	s.state.Unlock()
+
+	sort.Slice(tasks, func(l, r int) bool {
+		return tasks[l].Kind() < tasks[r].Kind()
+	})
+
+	// since device-service.access is unset, then both tasks should be queued
+	c.Assert(tasks, HasLen, 2)
+	c.Check(tasks[0].Kind(), Equals, "generate-device-key")
+	c.Check(tasks[1].Kind(), Equals, "request-serial")
+}
+
+func (s *deviceMgrSerialSuite) TestDeviceManagerNoAccessHasKeyID(c *C) {
+	s.state.Lock()
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Brand: "canonical",
+		Model: "pc",
+		KeyID: "key-id",
+	})
+
+	s.makeModelAssertionInState(c, "canonical", "pc", map[string]interface{}{
+		"architecture": "amd64",
+		"kernel":       "pc-kernel",
+		"gadget":       "pc",
+	})
+
+	devicestatetest.MockGadget(c, s.state, "pc", snap.R(2), nil)
+	s.state.Set("seeded", true)
+
+	tr := config.NewTransaction(s.state)
+	err := tr.Set("pc", "device-service.access", "offline")
+	c.Assert(err, IsNil)
+	tr.Commit()
+
+	s.state.Unlock()
+
+	err = s.mgr.Ensure()
+	c.Assert(err, IsNil)
+
+	s.state.Lock()
+	tasks := s.state.Tasks()
```

Comment:
Could you explain the reason to use one over the other, when you get a chance? Is it just so that the tests are more isolated, and something else adding a task would not break them?